### PR TITLE
Proposed extensible bitwise enum for file attributes.

### PR DESCRIPTION
**Pull request by katmsft**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp [hunk 1]
```diff
@@ -2,218 +2,82 @@
...
\ No newline at end of file
+namespace Azure { namespace Storage { namespace Files { namespace Shares { namespace Models {
+  class FileAttributes {
+  public:
+    FileAttributes() = default;
+    explicit FileAttributes(std::string value);
+    bool operator==(const FileAttributes& other) const;
+    bool operator!=(const FileAttributes& other) const;
+    FileAttributes operator|(FileAttributes rhs);
+    FileAttributes operator&(FileAttributes rhs);
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]: also `operator |=`

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]: maybe `operator ^`?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp [hunk 2]
```diff
@@ -2,218 +2,82 @@
...
-    }
-  } // namespace Details
-
-}}}} // namespace Azure::Storage::Files::Shares
\ No newline at end of file
+namespace Azure { namespace Storage { namespace Files { namespace Shares { namespace Models {
+  class FileAttributes {
+  public:
+    FileAttributes() = default;
+    explicit FileAttributes(std::string value);
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:13]:
```suggestion
    explicit FileAttributes(const std::string& value);
```

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp (hunk 1 above)
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]:
```suggestion
    FileAttributes operator&(const FileAttributes& rhs);
```

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]: What's your consideration of making it member function over non-member function?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp [hunk 3]
```diff
@@ -2,218 +2,82 @@
...
+    bool operator==(const FileAttributes& other) const;
+    bool operator!=(const FileAttributes& other) const;
+    FileAttributes operator|(FileAttributes rhs);
+    FileAttributes operator&(FileAttributes rhs);
+    const std::string& Get();
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:18]: Should this return a `std::string` or `std::vector<std::string>`

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:18]: Or do we need both functions, one returning `std::string`, one returning `std::vector<std::string>`.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]: what if the user wants to remove a specific attribute?

Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]: Are you suggesting we also support operator `-`?

Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]: I think there is no benefit for making it non-member function, given that we don't want implicit conversions for them. Keeping it in class is just more neat.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp (hunk 3 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:18]: I do think we need the `std::string` return value which is used when sending the request. I don't know when getting the vector would be in use.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:18]: What if a new symbol is added in the future, but our client sdk hasn't supported it yet. Now the user needs to be able to get the new symbol, right? This is the original intention of extensible enum. How is the user gonna do that?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:17]: > Are you suggesting we also support operator `-`?

No. I'm talking about `operator ^`

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp (hunk 3 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_file_attribute.hpp:18]: Makes sense.

**Comment by ahsonkhan** [comment]
@katmsft what's the plan for this PR? This will be a breaking change, so will need a CL entry.

**Pull request closed without merging**